src/tir/transforms/inject_software_pipeline.cc

```diff
@@ -0,0 +1,785 @@
+  }
+
+  arith::Analyzer analyzer_;
+  Map<Var, Buffer> buffer_data_to_buffer_;
+  const std::unordered_set<Buffer, ObjectPtrHash, ObjectPtrEqual>& double_buffers_;
+  Array<Buffer> pipeline_allocs_;
+  For pipeline_loop_;
+  PipelineInfo pipeline_info_;
+  const std::unordered_map<const VarNode*, FragmentInfo>& fragment_info_;
+  int max_stage_ = -1;
+  Map<Buffer, Buffer> buffer_remap_;
+  Array<Block> ordered_stmts_;
+};
+
+class PipelineInjector : private StmtExprMutator {
+ public:
+  static Stmt Inject(const PrimFunc& func) {
+    PipelineInjector injector;
+    for (const auto& kv : func->buffer_map) {
+      const Buffer& buffer = kv.second;
+      injector.buffer_data_to_buffer_.Set(buffer->data, buffer);
+    }
+    injector.fragment_info_ = GetTensorCoreFragmentInfo(func->body);
+    return injector(func->body);
+  }
+
+ private:
+  PipelineInjector() = default;
+
+  /*!
+   * \brief Check the pipeline satisfies the following conditions:
+   * 1) No conflicting order: The order of each statement should be unique.
+   * 2) No reordering with the same stage: Statements in the same stage are not allowed to be
+   * reordered.
+   */
+  void ValidatePipelineBody(const PipelineInfo& pipeline_info, const Array<Block>& original_order) {
+    std::unordered_set<int> used_orders;
+    std::unordered_map<int, int> stage_max_order;
+    for (const Block& block : original_order) {
+      const auto& stmt_info = pipeline_info.at(block);
+      int stage = stmt_info.stage;
+      int order = stmt_info.order;
+      CHECK(!used_orders.count(order))
+          << "ValueError: Two statements in the software pipeline cannot have the same order";
+      used_orders.insert(order);
+      CHECK(!stage_max_order.count(stage) || stage_max_order[stage] < order)
+          << "ValueError: Statements in the same stage of the software pipeline must have "
+             "increasing order.";
+      stage_max_order[stage] = order;
+    }
+  }
+
+  Stmt VisitStmt_(const ForNode* op) final {
+    // Step 1: Recursively rewrite the children first.
+    For for_node = Downcast<For>(StmtExprMutator::VisitStmt_(op));
+    bool is_pipeline = HasPipelineAnnotation(op);
+    if (!is_pipeline) {
+      return std::move(for_node);
+    }
+    // Step 2: Find the body and buffer allocations of the pipeline. The body can be direct child of
+    // the for-loop. If the for-loop has BlockRealize as its child, the pipeline body will be the
+    // child of the block.
+    Stmt pipeline_body{nullptr};
+    Array<Buffer> pipeline_allocs;
+    if (const auto* realize = for_node->body.as<BlockRealizeNode>()) {
+      const auto& block = realize->block;
+      for (const auto& buffer : block->alloc_buffers) {
+        ICHECK(buffer->IsInstance<BufferNode>());
+        buffer_data_to_buffer_.Set(buffer->data, buffer);
+      }
+      pipeline_body = block->body;
+      pipeline_allocs = block->alloc_buffers;
+    } else {
+      pipeline_body = for_node->body;
+    }
```

Comment:
I'm hitting an unexpected behavior where the pipeline transform generates functionally incorrect code.

If the annotated `for` loop _does not_ have a block as its child, but _does_ have an intermediate buffer allocation, that buffer will not be rewritten by the transform. The rest of the transform proceeds normally. The TIR can be generated and executed without any errors, but the results will be incorrect (because the intermediate buffer is not duplicated).

Here's a small example that exhibits the issue:
```python
@T.prim_func
def missing_block_example(A: T.Buffer[(16,), "float32"], C: T.Buffer[(16,), "float32"]):
    T.func_attr({"global_symbol": "main"})
    for i in T.serial(
            0, 16, annotations={"software_pipeline_stage": [0, 1], "software_pipeline_order": [0, 1]}
    ):
        #with T.block():  # <-- with this missing, the pipeline transform generates executable but incorrect code                                                                                                                                                                                                                                                                                                                                                                                                                    
            B = T.alloc_buffer((1,), "float32", scope="shared")
            with T.block():
                B[0] = A[i]
            with T.block():
                C[i] = B[0] + T.float32(1)
```

If you remove the `#` to include the `T.block` in the example, it works normally:
```
@main = primfn(A_handle: handle, C_handle: handle) -> ()
  attr = {"global_symbol": "main"}
  buffers = {A: Buffer(A_1: Pointer(global float32), float32, [16], []),
             C: Buffer(C_1: Pointer(global float32), float32, [16], [])}
  buffer_map = {C_handle: C, A_handle: A} {
  block([], "root") {
    tir.reads([])
    tir.writes([])
    block([], "") {
      tir.reads([A[0:16]])
      tir.writes([C[0:16]])
      B = alloc_buffer(float32[2, 1])
       {
        block([], "") {
          tir.reads([A[0]])
          tir.writes([B[0, 0]])
          B[0, 0] = A[0]
        block([], "") {
          tir.reads([A[1:16], B[0:2, 0]])
          tir.writes([B[0:2, 0], C[0:15]])
          for (i: int32, 0, 15) {
            block([], "") {
              tir.reads([A[(i + 1)]])
              tir.writes([B[floormod((i + 1), 2), 0]])
              B[floormod((i + 1), 2), 0] = A[(i + 1)]
            block([], "") {
              tir.reads([B[floormod(i, 2), 0]])
              tir.writes([C[i]])
              C[i] = (B[floormod(i, 2), 0] + 1f32)
          }
        block([], "") {
          tir.reads([B[1, 0]])
          tir.writes([C[15]])
          C[15] = (B[1, 0] + 1f32)
      }
}
```

With the `block` commented out (as above), it produces incorrect TIR:
```
@main = primfn(A_handle: handle, C_handle: handle) -> ()
  attr = {"global_symbol": "main"}
  buffers = {A: Buffer(A_1: Pointer(global float32), float32, [16], []),
             C: Buffer(C_1: Pointer(global float32), float32, [16], [])}
  buffer_map = {C_handle: C, A_handle: A} {
  block([], "root") {
    tir.reads([])
    tir.writes([])
    B = alloc_buffer(float32[1])
    block([], "") {
      tir.reads([A[0:16], B[0]])
      tir.writes([B[0], C[0:16]])
       {
        block([], "") {
          tir.reads([A[0]])
          tir.writes([B[0]])
          B[0] = A[0]
        block([], "") {
          tir.reads([A[1:16], B[0]])
          tir.writes([B[0], C[0:15]])
          for (i: int32, 0, 15) {
            block([], "") {
              tir.reads([A[(i + 1)]])
              tir.writes([B[0]])
              B[0] = A[(i + 1)]
            block([], "") {
              tir.reads([B[0]])
              tir.writes([C[i]])
              C[i] = (B[0] + 1f32)
          }
        block([], "") {
          tir.reads([B[0]])
          tir.writes([C[15]])
          C[15] = (B[0] + 1f32)
      }
}
```

(In the first output, buffer `B` is rewritten and has all accesses to it rewritten. In the second, it is not changed at all.)

I'm wondering if it's "incorrect" TIR to have a buffer allocate as a child of a `for` loop like that, but I don't get any errors and can build and execute the program fine.

Reply:
This is a great question. `alloc_buffer` has to be under a block. So the example provided is not a correct TIR, if you print the above TIR, the result will be
```
# from tvm.script import tir as T
@T.prim_func
def func(A: T.Buffer[(16,), "float32"], C: T.Buffer[(16,), "float32"]) -> None:
    # function attr dict
    T.func_attr({"global_symbol": "main"})
    # body
    # with T.block("root")
    B = T.alloc_buffer([1], dtype="float32", scope="shared")
    for i in T.serial(16, annotations={"software_pipeline_order":[0, 1], "software_pipeline_stage":[0, 1]}):
        with T.block():
            T.reads(A[i])
            T.writes(B[0])
            B[0] = A[i]
        with T.block():
            T.reads(B[0])
            T.writes(C[i])
            C[i] = B[0] + T.float32(1)
```
you can see that the parser actually move the `alloc_buffer` to the root block.

Besides, the pass runs after [`PlanAndUpdateBufferAllocationLocation`](https://github.com/apache/tvm/blob/main/src/driver/driver_api.cc#L245). This pass will compute the LCA of the buffer accessor, and insert a `BlockRealize`/`Block` to allocate the buffer.